Fix for date deserialization in Safari

kiwio opened the pull request:
There's a known behavior in Safari that applies current OS time zone to the date object when the Date constructor (or Date.parse method) is used. Because of this serializr will produce the date of 2017/12/07 in New Zealand time zone for the value of '2017-12-06T15:37:39.903'. The suggested fix will mitigate the issue for ISO 8601 format date strings without timezones.

See the following link for more details:
https://stackoverflow.com/a/33909265

@mweststrate

kiwio commented:
cc @alexggordon

alexggordon commented:
@kiwio, hey sorry, I've been pretty busy the last month. I'll see if I can get to this ASAP

alexggordon commented:
So after thinking about this, I'm wondering if instead of testing for just `typeof ... === "string"` if we should `&&` regex test for ISO-8601 validity also. The benefit being that if there's something wrong with the string (it's not in the correct format) the Date Library would raise the error as opposed to the stack trace being to use trying to access an array element. 

Looks like there's a pretty good regex [here](https://www.regextester.com/97766)

@kiwio thoughts?

kiwio commented:
Hi @alexggordon 

Thanks for your reply.

Unfortunately the suggested solution didn't work out for us and introduced more issues than it solved. The behavior of Date object turned out to be so disparate in different browsers so we had to approach this more radically and replaced dates completely with moment objects.

Here's the moment decorator we came up with, it might be worth considering including it into serializr library:
```
import * as momentjs from 'moment';

export function moment() {
    return {
        serializer: function (value: momentjs.Moment): any {
            if (value === null || value === undefined) {
                return value;
            }
            if (!momentjs.isMoment(value)) {
                throw new Error('[serializr] Expected Moment object');
            }
            return value.toISOString();
        },
        deserializer: function (value: any, done: Function): momentjs.Moment {
            if (value === null || value === undefined) {
                return void done(null, value);
            }
            const date = momentjs.utc(value);
            return void done(null, date);
        }
    };
}
```

alexggordon commented:
Hey @kiwio , thanks for the detailed and thorough reply, it really helped!
